drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,313 @@
+mod instr {
+    pub const PAGE_PROGRAM: u32 = 0x02;
+    pub const READ: u32 = 0x03;
+    pub const READ_STATUS_1: u32 = 0x05;
+    pub const READ_STATUS_2: u32 = 0x35;
+    pub const READ_STATUS_3: u32 = 0x15;
+    pub const WRITE_STATUS_2: u32 = 0x31;
+    pub const WRITE_ENABLE: u32 = 0x06;
+    pub const FAST_READ_QUAD: u32 = 0x6b;
+    pub const FAST_READ_QUAD_OUTPUT_4B: u32 = 0x6c;
+    pub const SECTOR_ERASE: u32 = 0x20;
+    pub const READ_JEDEC_ID: u32 = 0x9f;
+    pub const BLOCK_ERASE_64KB: u32 = 0xd8;
+    pub const BLOCK_ERASE_64KB_4B: u32 = 0xdc;
+    pub const QUAD_INPUT_PAGE_PROGRAM: u32 = 0x32;
+    pub const QUAD_INPUT_PAGE_PROGRAM_4B: u32 = 0x34;
+}
+
+impl ServerImpl {
+    fn flash_read_id(&self) -> [u8; 20] {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 20);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_JEDEC_ID);
+        self.wait_fpga_busy();
+        let mut out = [0u8; 20];
+        for i in 0..out.len() / 4 {
+            let v = self.read_reg(reg::RX_FIFO);
+            for (j, byte) in v.to_le_bytes().iter().enumerate() {
+                out[i * 4 + j] = *byte;
+            }
+        }
+        out
+    }
+
+    fn read_reg(&self, reg: *mut u32) -> u32 {
+        unsafe { reg.read_volatile() }
+    }
+
+    fn write_reg(&self, reg: *mut u32, v: u32) {
+        unsafe { reg.write_volatile(v) };
+    }
+
+    /// Wait until the FPGA is idle
+    fn wait_fpga_busy(&self) {
+        loop {
+            let spi_sr = self.read_reg(reg::SPISR);
+            if (spi_sr & 1) == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy { spi_sr });
+            sleep_for(1);
+        }
+    }
+
+    /// Wait until a word is available in the FPGA's RX buffer
+    fn wait_fpga_rx(&self) {
+        for i in 0.. {
+            let spi_sr = self.read_reg(reg::SPISR);
+            const RX_EMPTY_BIT: u32 = 1 << 6;
+            if spi_sr & RX_EMPTY_BIT == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy { spi_sr });
+            // Initial busy-loop for faster response
+            if i >= 32 {
+                sleep_for(1);
+            }
+        }
+    }
+
+    /// Clears the FPGA's internal FIFOs
+    fn clear_fifos(&self) {
+        self.write_reg(reg::SPICR, 0x8080);
+    }
+
+    /// Wait until the SPI flash is idle
+    fn wait_flash_busy(&self, t: Trace) {
+        // Wait for the busy flag to be unset
+        while (self.read_flash_status() & 1) != 0 {
+            ringbuf_entry!(t);
+            sleep_for(1);
+        }
+    }
+
+    /// Reads the STATUS1 register from flash
+    fn read_flash_status(&self) -> u8 {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 1);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_STATUS_1);
```

Comment:
there is a part of me that wonders if it's worth taking all this and putting it in a function that takes the number of bytes and the instruction value, but, honestly, I'm not sure if it would be worth doing, because you'd still have to provide the number of bytes, address, dummy cycles, and instructions to that function so i don't really think it would be abstracting over anything meaningful...and this feels more explicit, i guess?